src/file.rs

```diff
@@ -193,26 +138,36 @@ impl TableFile {
 		capacity += GROW_SIZE_BYTES / entry_size as u64;
 
 		self.capacity.store(capacity, Ordering::Relaxed);
-		let mut file = self.file.upgradable_read();
-		if file.is_none() {
-			let mut wfile = RwLockUpgradableReadGuard::upgrade(file);
-			*wfile = Some(self.create_file()?);
-			file = RwLockWriteGuard::downgrade_to_upgradable(wfile);
+		let mut map_and_file = self.map.write();
+		match map_and_file.as_mut() {
+			None => {
+				let file = self.create_file()?;
+				try_io!(file.set_len(capacity * entry_size as u64));
+				let mut map = try_io!(unsafe { memmap2::MmapMut::map_mut(&file) });
+				madvise_random(&mut map);
+				*map_and_file = Some((map, file));
+			},
+			Some((map, file)) => {
+				try_io!(file.set_len(capacity * entry_size as u64));
+				let mut m = try_io!(unsafe { memmap2::MmapMut::map_mut(&*file) });
+				madvise_random(&mut m);
+				*map = m;
+			},
 		}
-		try_io!(file.as_ref().unwrap().set_len(capacity * entry_size as u64));
 		Ok(())
 	}
 
 	pub fn flush(&self) -> Result<()> {
-		if let Some(file) = self.file.read().as_ref() {
```

Comment:
in https://github.com/cheme/parity-db/blob/3b49e8724f30573570faab9f8902ad7b65211bc1/src/file.rs#L470 I did use flush range to avoid flushing the start of the file (probably the header), but tbh it does not really make sense to me right now, just mentioning in case there may be some use to it.

Reply:
Header for a value table is enacted as a regular entry, so it should be synced as well.